File: frontend/src/types/finanzas.ts
export type TipoIngreso = 'pago_cliente' | 'anticipo' | 'finiquito' | 'otros'
export type CategoriaEgreso = 'materiales' | 'mano_obra' | 'subcontratistas' | 'maquinaria' | 'gastos_generales' | 'impuestos' | 'otros'
export type MetodoPago = 'transferencia' | 'cheque' | 'efectivo' | 'tarjeta'
export type EstadoPago = 'pendiente' | 'pagado' | 'parcial'
export type EstadoCuenta = 'pendiente' | 'parcial' | 'pagada' | 'vencida'

// Labels y colores para UI
export const tipoIngresoLabels: Record<TipoIngreso, string> = {
  pago_cliente: 'Pago Cliente',
  anticipo: 'Anticipo',
  finiquito: 'Finiquito',
  otros: 'Otros',
}

export const tipoIngresoColors: Record<TipoIngreso, string> = {
  pago_cliente: 'success',
  anticipo: 'info',
  finiquito: 'purple',
  otros: 'secondary',
}

export const categoriaEgresoLabels: Record<CategoriaEgreso, string> = {
  materiales: 'Materiales',
  mano_obra: 'Mano de Obra',
  subcontratistas: 'Subcontratistas',
  maquinaria: 'Maquinaria',
  gastos_generales: 'Gastos Generales',
  impuestos: 'Impuestos',
  otros: 'Otros',
}

export const categoriaEgresoColors: Record<CategoriaEgreso, string> = {
  materiales: 'info',
  mano_obra: 'orange',
  subcontratistas: 'purple',
  maquinaria: 'cyan',
  gastos_generales: 'secondary',
  impuestos: 'warning',
  otros: 'secondary',
}

export const metodoPagoLabels: Record<MetodoPago, string> = {
  transferencia: 'Transferencia',
  cheque: 'Cheque',
  efectivo: 'Efectivo',
  tarjeta: 'Tarjeta',
}

export const estadoPagoLabels: Record<EstadoPago, string> = {
  pendiente: 'Pendiente',
  pagado: 'Pagado',
  parcial: 'Parcial',
}

export const estadoPagoColors: Record<EstadoPago, string> = {
  pendiente: 'warning',
  pagado: 'success',
  parcial: 'info',
}

export const estadoCuentaLabels: Record<EstadoCuenta, string> = {
  pendiente: 'Pendiente',
  parcial: 'Parcial',
  pagada: 'Pagada',
  vencida: 'Vencida',
}

export const estadoCuentaColors: Record<EstadoCuenta, string> = {
  pendiente: 'warning',
  parcial: 'info',
  pagada: 'success',
  vencida: 'error',
}

export interface Ingreso {
  id: string
  tipo: TipoIngreso
  concepto: string
  monto: number
  fecha: string
  proyecto_id: string | null
  cliente_id: string | null
  metodo_pago: MetodoPago | null
  referencia: string | null
  factura: string | null
  notas: string | null
  registrado_por_id: string
  activo: boolean
  created_at: string
}

export interface IngresoCreate {
  tipo: TipoIngreso
  concepto: string
  monto: number
  fecha: string
  proyecto_id?: string
  cliente_id?: string
  metodo_pago?: MetodoPago
  referencia?: string
  factura?: string
  notas?: string
}

export interface Egreso {
  id: string
  categoria: CategoriaEgreso
  subcategoria: string | null
  concepto: string
  monto: number
  fecha: string
  proyecto_id: string | null
  proveedor_id: string | null
  subcontratista_id: string | null
  empleado_id: string | null
  metodo_pago: MetodoPago | null
  referencia: string | null
  factura: string | null
  autorizado_por_id: string | null
  estado_pago: EstadoPago
  notas: string | null
  activo: boolean
  created_at: string
}

export interface EgresoCreate {
  categoria: CategoriaEgreso
  subcategoria?: string
  concepto: string
  monto: number
  fecha: string
  proyecto_id?: string
  proveedor_id?: string
  subcontratista_id?: string
  empleado_id?: string
  metodo_pago?: MetodoPago
  referencia?: string
  factura?: string
  notas?: string
}

export interface PagoCuenta {
  id: string
  monto: number
  fecha: string
  metodo_pago: MetodoPago | null
  referencia: string | null
  notas: string | null
}

export interface CuentaPorCobrar {
  id: string
  cliente_id: string
  proyecto_id: string | null
  concepto: string
  monto_total: number
  monto_pagado: number
  saldo: number
  fecha_emision: string
  fecha_vencimiento: string | null
  estado: string
  pagos: PagoCuenta[]
  activo: boolean
  created_at: string
}

export interface CuentaCobrarCreate {
  cliente_id: string
  proyecto_id?: string
  concepto: string
  monto_total: number
  fecha_emision: string
  fecha_vencimiento?: string
}

export interface CuentaPorPagar {
  id: string
  proveedor_id: string | null
  subcontratista_id: string | null
  proyecto_id: string | null
  concepto: string
  monto_total: number
  monto_pagado: number
  saldo: number
  fecha_emision: string
  fecha_vencimiento: string | null
  estado: string
  prioridad: string
  pagos: PagoCuenta[]
  activo: boolean
  created_at: string
}

export interface CuentaPagearCreate {
  proveedor_id?: string
  subcontratista_id?: string
  proyecto_id?: string
  concepto: string
  monto_total: number
  fecha_emision: string
  fecha_vencimiento?: string
  prioridad?: string
}

export interface PagoCreate {
  monto: number
  fecha: string
  metodo_pago?: MetodoPago
  referencia?: string
  notas?: string
}

File: backend/app/models/finanzas.py
import enum
from sqlalchemy import Column, String, Integer, Numeric, Date, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import relationship
from app.models.base import BaseModel


class TipoIngreso(str, enum.Enum):
    PAGO_CLIENTE = "pago_cliente"
    ANTICIPO = "anticipo"
    FINIQUITO = "finiquito"
    OTROS = "otros"


class CategoriaEgreso(str, enum.Enum):
    MATERIALES = "materiales"
    MANO_OBRA = "mano_obra"
    SUBCONTRATISTAS = "subcontratistas"
    MAQUINARIA = "maquinaria"
    GASTOS_GENERALES = "gastos_generales"
    IMPUESTOS = "impuestos"
    OTROS = "otros"


class MetodoPago(str, enum.Enum):
    TRANSFERENCIA = "transferencia"
    CHEQUE = "cheque"
    EFECTIVO = "efectivo"
    TARJETA = "tarjeta"


class EstadoPago(str, enum.Enum):
    PENDIENTE = "pendiente"
    PAGADO = "pagado"
    PARCIAL = "parcial"


class Ingreso(BaseModel):
    __tablename__ = "ingresos"

    tipo = Column(SAEnum(TipoIngreso), nullable=False)
    concepto = Column(String(255), nullable=False)
    monto = Column(Numeric(15, 2), nullable=False)
    fecha = Column(Date, nullable=False)

    proyecto_id = Column(String(36), ForeignKey("proyectos.id"), nullable=True)
    cliente_id = Column(String(36), ForeignKey("clientes.id"), nullable=True)

    metodo_pago = Column(SAEnum(MetodoPago), nullable=True)
    referencia = Column(String(100), nullable=True)
    factura = Column(String(100), nullable=True)
    comprobante_url = Column(String(500), nullable=True)

    notas = Column(String, nullable=True)
    registrado_por_id = Column(String(36), ForeignKey("usuarios.id"), nullable=False)


class Egreso(BaseModel):
    __tablename__ = "egresos"

    categoria = Column(SAEnum(CategoriaEgreso), nullable=False)
    subcategoria = Column(String(100), nullable=True)
    concepto = Column(String(255), nullable=False)
    monto = Column(Numeric(15, 2), nullable=False)
    fecha = Column(Date, nullable=False)

    proyecto_id = Column(String(36), ForeignKey("proyectos.id"), nullable=True)
    unidad_id = Column(String(36), ForeignKey("unidades.id"), nullable=True)  # PH asignado
    proveedor_id = Column(String(36), ForeignKey("proveedores.id"), nullable=True)
    subcontratista_id = Column(String(36), ForeignKey("subcontratistas.id"), nullable=True)
    empleado_id = Column(String(36), ForeignKey("empleados.id"), nullable=True)

    metodo_pago = Column(SAEnum(MetodoPago), nullable=True)
    referencia = Column(String(100), nullable=True)
    factura = Column(String(100), nullable=True)
    comprobante_url = Column(String(500), nullable=True)

    autorizado_por_id = Column(String(36), ForeignKey("usuarios.id"), nullable=True)
    estado_pago = Column(SAEnum(EstadoPago), default=EstadoPago.PENDIENTE)

    notas = Column(String, nullable=True)

    # Relación con unidad
    unidad = relationship("Unidad", back_populates="egresos")


class CuentaPorCobrar(BaseModel):
    __tablename__ = "cuentas_por_cobrar"

    cliente_id = Column(String(36), ForeignKey("clientes.id"), nullable=False)
    proyecto_id = Column(String(36), ForeignKey("proyectos.id"), nullable=True)
    concepto = Column(String(255), nullable=False)
    monto_total = Column(Numeric(15, 2), nullable=False)
    monto_pagado = Column(Numeric(15, 2), default=0)
    saldo = Column(Numeric(15, 2), nullable=False)
    fecha_emision = Column(Date, nullable=False)
    fecha_vencimiento = Column(Date, nullable=True)
    estado = Column(String(30), default="pendiente")  # pendiente, parcial, pagada, vencida

    # Relaciones
    pagos = relationship("PagoCuentaCobrar", back_populates="cuenta", cascade="all, delete-orphan")


class PagoCuentaCobrar(BaseModel):
    __tablename__ = "pagos_cuenta_cobrar"

    cuenta_id = Column(String(36), ForeignKey("cuentas_por_cobrar.id"), nullable=False)
    monto = Column(Numeric(15, 2), nullable=False)
    fecha = Column(Date, nullable=False)
    metodo_pago = Column(SAEnum(MetodoPago), nullable=True)
    referencia = Column(String(100), nullable=True)
    notas = Column(String, nullable=True)

    # Relación
    cuenta = relationship("CuentaPorCobrar", back_populates="pagos")


class CuentaPorPagar(BaseModel):
    __tablename__ = "cuentas_por_pagar"

    proveedor_id = Column(String(36), ForeignKey("proveedores.id"), nullable=True)
    subcontratista_id = Column(String(36), ForeignKey("subcontratistas.id"), nullable=True)
    proyecto_id = Column(String(36), ForeignKey("proyectos.id"), nullable=True)
    concepto = Column(String(255), nullable=False)
    monto_total = Column(Numeric(15, 2), nullable=False)
    monto_pagado = Column(Numeric(15, 2), default=0)
    saldo = Column(Numeric(15, 2), nullable=False)
    fecha_emision = Column(Date, nullable=False)
    fecha_vencimiento = Column(Date, nullable=True)
    estado = Column(String(30), default="pendiente")  # pendiente, parcial, pagada, vencida
    prioridad = Column(String(10), default="media")  # baja, media, alta

    # Relaciones
    pagos = relationship("PagoCuentaPagar", back_populates="cuenta", cascade="all, delete-orphan")


class PagoCuentaPagar(BaseModel):
    __tablename__ = "pagos_cuenta_pagar"

    cuenta_id = Column(String(36), ForeignKey("cuentas_por_pagar.id"), nullable=False)
    monto = Column(Numeric(15, 2), nullable=False)
    fecha = Column(Date, nullable=False)
    metodo_pago = Column(SAEnum(MetodoPago), nullable=True)
    referencia = Column(String(100), nullable=True)
    notas = Column(String, nullable=True)

    # Relación
    cuenta = relationship("CuentaPorPagar", back_populates="pagos")
